### PR TITLE
update `readonly` with new usages in new C# versions.

### LINKED ISSUE #5682: Please update with C# 7.2 usage
Please update with C# 7.2 ref readonly returns.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 26899b81-d6f9-534a-83ee-283dd8aae545
* Version Independent ID: 3061125d-7de4-97ef-ccb0-ee792d6f2ae1
* Content: [readonly (C# Reference)](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/readonly?f1url=https%3A%2F%2Fmsdn.microsoft.com%2Fquery%2Fdev15.query%3FappId%3DDev15IDEF1%26l%3DEN-US%26k%3Dk(readonly_CSharpKeyword)%3Bk(SolutionItemsProject)%3Bk(DevLang-csharp)%26rd%3Dtrue)
* Content Source: [docs/csharp/language-reference/keywords/readonly.md](https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/keywords/readonly.md)
* Service: **unspecified**
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #5682 

/cc @fitdev

Relies on dotnet/samples#133

### CODE REVIEW

**Comment by BillWagner** [comment]
closing and reopening to force a build.

**Comment by BillWagner** [comment]
yet another build.

File: docs/csharp/language-reference/keywords/readonly.md [hunk 1]
```diff
@@ -9,50 +9,86 @@ helpviewer_keywords:
...
+  
+which is the same error you get when you attempt to assign a value to a constant.  
+
+## Readonly struct example
+
+The `readonly` modifier on a `struct` definition declares that the struct is **immutable**. Every instance field of the `struct` must be marked `readonly`, as shown in the following example:
+
+[!code-csharp[readonly struct example](~/samples/snippets/csharp/keywords/ReadonlyKeywordExamples.cs#ReadonlyStruct)]  
+
+The preceding example uses readonly auto properties to declare its storage. That instructs the compiler to create `readonly` backing fields for those properties. You could also declare `readonly` fields directly:
```
Comment (JRAlexander) [inline docs/csharp/language-reference/keywords/readonly.md]: Should there be a link to readonly auto properties?

File: docs/csharp/language-reference/keywords/readonly.md [hunk 2]
```diff
@@ -9,50 +9,86 @@ helpviewer_keywords:
...
+You can assign a value to a `readonly` field only in the following contexts:  
   
--   For an instance field, in the instance constructors of the class that contains the field declaration, or for a static field, in the static constructor of the class that contains the field declaration. These are also the only contexts in which it is valid to pass a `readonly` field as an [out](../../../csharp/language-reference/keywords/out-parameter-modifier.md) or [ref](../../../csharp/language-reference/keywords/ref.md) parameter.  
+- When the variable is initialized in the declaration, for example:  
+
+```csharp
+public readonly int y = 5;  
+```
+
+- For an instance field, in the instance constructors of the class that contains the field declaration, or for a static field, in the static constructor of the class that contains the field declaration. These are also the only contexts in which it is valid to pass a `readonly` field as an [out](out-parameter-modifier.md) or [ref](ref.md) parameter.
```
Comment (JRAlexander) [inline docs/csharp/language-reference/keywords/readonly.md]: This first sentence is kind of awkward. I'm not sure what you are trying to say.

File: docs/csharp/language-reference/keywords/readonly.md [hunk 3]
```diff
@@ -9,50 +9,86 @@ helpviewer_keywords:
...
+  
+you will get the compiler error message:  
+  
+`The left-hand side of an assignment must be an l-value`  
+  
+which is the same error you get when you attempt to assign a value to a constant.  
+
+## Readonly struct example
+
+The `readonly` modifier on a `struct` definition declares that the struct is **immutable**. Every instance field of the `struct` must be marked `readonly`, as shown in the following example:
```
Comment (JRAlexander) [inline docs/csharp/language-reference/keywords/readonly.md]: Should immutable be surrounded by "`"?

**Review by JRAlexander — CHANGES_REQUESTED** [review]
Bill, looks good , except for a few small items.

File: docs/csharp/language-reference/keywords/readonly.md (hunk 3 above)
Comment (BillWagner) [reply docs/csharp/language-reference/keywords/readonly.md]: Nope, that's not a keyword.

**Comment by BillWagner** [comment]
Thanks @JRAlexander 
Changes have been addressed.

**Review by JRAlexander — APPROVED** [review]

**Comment by JRAlexander** [comment]
LGTM. :shipit: when ready!

**Pull request merged**